File: src/groklang/lexer.py
import ply.lex as lex

# Token list
tokens = (
    'INT', 'FLOAT', 'STRING', 'RAW_STRING', 'BYTE_STRING', 'CHAR', 'TRUE', 'FALSE', 'ID', 'UNDERSCORE',
    'FN', 'LET', 'MUT', 'IF', 'ELSE', 'MATCH', 'FOR', 'WHILE', 'LOOP', 'IN',
    'BREAK', 'CONTINUE', 'RETURN', 'STRUCT', 'ENUM', 'TRAIT', 'IMPL', 'USE',
    'MOD', 'PUB', 'SELF', 'SELFTYPE', 'WHERE', 'TYPE', 'AS', 'UNSAFE',
    'EXTERN', 'CONST', 'STATIC', 'MACRO', 'AI', 'ACTOR', 'SPAWN', 'MOVE',
    # Operators...
    'PLUS', 'MINUS', 'TIMES', 'DIVIDE', 'MODULO',
    'EQ', 'NE', 'LT', 'GT', 'LE', 'GE',
    'AND', 'OR', 'NOT', 'AMPERSAND', 'PIPE', 'CARET', 'TILDE',
    'LSHIFT', 'RSHIFT', 'ASSIGN',
    'LPAREN', 'RPAREN', 'LBRACE', 'RBRACE', 'LBRACKET', 'RBRACKET',
    'COMMA', 'SEMICOLON', 'COLON', 'DOUBLECOLON', 'DOT', 'ARROW', 'FATARROW',
    'ELLIPSIS', 'QUESTION', 'AT', 'HASH', 'DOLLAR',
)

reserved = {
    'fn': 'FN', 'let': 'LET', 'mut': 'MUT', 'if': 'IF',
    'else': 'ELSE', 'match': 'MATCH', 'for': 'FOR', 'while': 'WHILE',
    'loop': 'LOOP', 'break': 'BREAK', 'continue': 'CONTINUE',
    'return': 'RETURN', 'struct': 'STRUCT', 'enum': 'ENUM',
    'trait': 'TRAIT', 'impl': 'IMPL', 'use': 'USE',
    'mod': 'MOD', 'pub': 'PUB', 'self': 'SELF', 'Self': 'SELFTYPE',
    'true': 'TRUE', 'false': 'FALSE', 'where': 'WHERE',
    'type': 'TYPE', 'as': 'AS', 'unsafe': 'UNSAFE',
    'extern': 'EXTERN', 'const': 'CONST', 'static': 'STATIC',
    'macro': 'MACRO', 'macro_rules': 'MACRO_RULES', 'ai': 'AI', 'actor': 'ACTOR', 'spawn': 'SPAWN',
    'move': 'MOVE', 'in': 'IN', '_': 'UNDERSCORE',
}

# Token rules
def t_RAW_STRING(t):
    r'r"([^"])*"'
    t.value = t.value[2:-1]  # Remove r"
    return t

def t_BYTE_STRING(t):
    r'b"([^"\\\\]|\\\\.)*"'
    t.value = t.value[2:-1]  # Remove b"
    t.value = t.value.encode('utf-8')  # Convert to bytes
    return t

def t_ID(t):
    r'[a-zA-Z_][a-zA-Z_0-9]*'
    t.type = reserved.get(t.value, 'ID')
    return t

def t_FLOAT(t):
    r'(\d+\.\d+|\d*\.\d+)([eE][+-]?\d+)?[fF]?|(\d+)[eE][+-]?\d+[fF]?'
    t.value = float(t.value)
    return t

def t_INT(t):
    r'(0[xX][0-9a-fA-F]+|0[bB][01]+|0[oO][0-7]+|\d+)[uUiI]?(8|16|32|64|128)?'
    # Strip suffix before converting
    value_str = t.value
    if any(c in value_str for c in 'uUiI'):
        # Find the start of suffix
        for i, c in enumerate(value_str):
            if c in 'uUiI':
                value_str = value_str[:i]
                break
    t.value = int(value_str, 0)
    return t

def t_STRING(t):
    r'"([^"\\\\]|\\\\.)*"'
    t.value = t.value[1:-1]
    return t

def t_CHAR(t):
    r"'([^'\\\\]|\\\\.)?'"
    t.value = t.value[1:-1]
    return t

# Operators (longest match first)
t_EQ = r'=='
t_NE = r'!='
t_LE = r'<='
t_GE = r'>='
t_AND = r'&&'
t_OR = r'\|\|'
t_LSHIFT = r'<<'
t_RSHIFT = r'>>'
t_DOUBLECOLON = r'::'
t_ARROW = r'->'
t_FATARROW = r'=>'
t_ELLIPSIS = r'\.\.\.'

t_PLUS = r'\+'
t_MINUS = r'-'
t_TIMES = r'\*'
t_DIVIDE = r'/'
t_MODULO = r'%'
t_LT = r'<'
t_GT = r'>'
t_NOT = r'!'
t_AMPERSAND = r'&'
t_PIPE = r'\|'
t_CARET = r'\^'
t_TILDE = r'~'
t_ASSIGN = r'='
t_LPAREN = r'\('
t_RPAREN = r'\)'
t_LBRACE = r'\{'
t_RBRACE = r'\}'
t_LBRACKET = r'\['
t_RBRACKET = r'\]'
t_COMMA = r','
t_SEMICOLON = r';'
t_COLON = r':'
t_DOT = r'\.'
t_QUESTION = r'\?'
t_AT = r'@'
t_DOLLAR = r'\$'

# Comments
def t_COMMENT(t):
    r'//[^\n]*'
    pass

def t_MCOMMENT(t):
    r'/\*(.|\n)*?\*/'
    t.lexer.lineno += t.value.count('\n')
    pass

def t_newline(t):
    r'\n+'
    t.lexer.lineno += len(t.value)

t_ignore = ' \t\r'

def t_error(t):
    print(f"Illegal character '{t.value[0]}' at line {t.lexer.lineno}")
    t.lexer.skip(1)

lexer = lex.lex(debug=0, optimize=0)

File: grok/src/lexer.rs
use logos::Logos;

#[derive(Logos, Debug, PartialEq, Clone)]
#[logos(skip r"[ \t\n\f]+")]
pub enum Token {
    #[token("fn")]
    Fn,
    #[token("let")]
    Let,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("match")]
    Match,
    #[token("case")]
    Case,
    #[token("async")]
    Async,
    #[token("await")]
    Await,
    #[token("actor")]
    Actor,
    #[token("spawn")]
    Spawn,
    #[token("send")]
    Send,
    #[token("receive")]
    Receive,
    #[token("module")]
    Module,
    #[token("macro")]
    Macro,
    #[token("macro_rules")]
    MacroRules,
    #[token("ai")]
    Ai,
    #[token("test")]
    Test,
    #[token("optimize")]
    Optimize,
    #[token("pub")]
    Pub,
    #[token("mut")]
    Mut,
    #[token("const")]
    Const,
    #[token("import")]
    Import,
    #[token("as")]
    As,
    #[token("from")]
    From,
    #[token("struct")]
    Struct,
    #[token("enum")]
    Enum,
    #[token("trait")]
    Trait,
    #[token("impl")]
    Impl,
    #[token("use")]
    Use,
    #[token("mod")]
    Mod,
    #[token("self")]
    Self_,
    #[token("Self")]
    SelfType,
    #[token("true")]
    True,
    #[token("false")]
    False,
    #[token("where")]
    Where,
    #[token("type")]
    Type,
    #[token("unsafe")]
    Unsafe,
    #[token("extern")]
    Extern,
    #[token("static")]
    Static,
    #[token("move")]
    Move,
    #[token("in")]
    In,
    #[token("return")]
    Return,
    #[token("break")]
    Break,
    #[token("continue")]
    Continue,
    #[token("loop")]
    Loop,
    #[token("for")]
    For,
    #[token("while")]
    While,
    #[token("_", priority = 2)]
    Underscore,

    // Literals
    #[regex(r#""([^"\\]|\\.)*""#)]
    String,
    #[regex(r#"'([^'\\]|\\.)'"#)]
    Char,
    #[regex(r#"b"([^"\\]|\\.)*""#)]
    ByteString,
    #[regex(r"[0-9]+")]
    Int,
    #[regex(r"[0-9]+\.[0-9]+", priority = 2)]
    Float,

    // Operators
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Star,
    #[token("/")]
    Slash,
    #[token("%")]
    Percent,
    #[token("=")]
    Assign,
    #[token("==")]
    Eq,
    #[token("!=")]
    Ne,
    #[token("<")]
    Lt,
    #[token(">")]
    Gt,
    #[token("<=")]
    Le,
    #[token(">=")]
    Ge,
    #[token("&&")]
    And,
    #[token("||")]
    Or,
    #[token("!")]
    Bang,
    #[token("&")]
    Amp,
    #[token("|")]
    Pipe,
    #[token("^")]
    Caret,
    #[token("~")]
    Tilde,
    #[token("<<")]
    Shl,
    #[token(">>")]
    Shr,
    #[token("..")]
    DotDot,
    #[token("...")]
    DotDotDot,
    #[token("?")]
    Question,
    #[token("@")]
    At,
    #[token("#")]
    Hash,
    #[token("$")]
    Dollar,
    #[token("=>")]
    FatArrow,
    #[token("->")]
    Arrow,

    // Delimiters
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[token(":")]
    Colon,
    #[token("::")]
    DoubleColon,
    #[token(".")]
    Dot,

    // Identifiers
    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*", priority = 1)]
    Identifier,
}

pub struct TokenData {
    pub token: Token,
    pub slice: String,
    pub line: usize,
    pub col: usize,
}

pub struct Lexer<'source> {
    inner: logos::Lexer<'source, Token>,
    line: usize,
    last_newline: usize,
}

impl<'source> Lexer<'source> {
    pub fn new(source: &'source str) -> Self {
        Self {
            inner: Token::lexer(source),
            line: 1,
            last_newline: 0,
        }
    }
}

impl<'source> Iterator for Lexer<'source> {
    type Item = Result<TokenData, ()>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.inner.next() {
            Some(Ok(token)) => {
                let slice = self.inner.slice();
                let span = self.inner.span();
                
                // Update line and column
                let start = span.start;
                let col = start - self.last_newline + 1;
                
                // If the slice contains newlines, update our tracking for the NEXT token
                for (i, c) in slice.char_indices() {
                    if c == '\n' {
                        self.line += 1;
                        self.last_newline = start + i + 1;
                    }
                }

                Some(Ok(TokenData {
                    token,
                    slice: slice.to_string(),
                    line: self.line,
                    col,
                }))
            }
            Some(Err(_)) => Some(Err(())),
            None => None,
        }
    }
}
